napari/utils/_tests/test_io.py

```diff
@@ -123,6 +123,19 @@ def test_zarr():
         np.testing.assert_array_equal(image, image_in)
 
 
+@pytest.mark.skipif(not zarr_available, reason='zarr not installed')
+def test_zarr_nested(tmp_path):
+    image = np.random.random((10, 20, 20))
+    image_name = 'my_image'
+    root_path = tmp_path / 'dataset.zarr'
```

Comment:
Does this syntax `tmp_path / 'dataset.zarr'` automatically do a string concatenation with `/` as a file separator? Is there a problem with `/` vs `\` if it is a file separator across platforms? (evidently not as tests run, but just asking to best understand - apologies for my ignorance here!)

Reply:
That’s a (very cool) feature of pathlib.Path objects (which the tmp_path is): https://stackoverflow.com/questions/53083963/python-pathlib-operator-how-does-it-do-it
It’s not a string, but rather an operation, and it works across platforms like os.path.join

Reply:
Yes my understanding is that [`pathlib.Path`](https://docs.python.org/3/library/pathlib.html) implements `__truediv__` and deals with both separators. Zarr (if I remember correctly) cannot take a `pathlib.Path` object and requires a string, which is why I cast down when creating/reading the array.

Somewhat of a side note, but related, I noticed the `magic_read` docstring says `filenames` is `pathlib.Path | [pathlib.Path]` but `filenames` is immediately cast to a string. The function then operates on these strings with `os.path`. I was tempted to remove the top-level casting to strings and use the `pathlib.Path`s (as I assume may have been intended?) to to glob files and build paths, only converting to strings when calling thirdparty readers. However, that was beyond the scope of this PR.

For instance, the current implementation for `guess_zarr_path` is somewhat of an antipattern (if I'm using that word correctly). It requires the recreation a `pathlib.Path` to check the parts, but this wouldn't be necessary if the `filename` was still a `pathlib.Path`.

```python
def guess_zarr_path(filename):
    return any(
        part.endswith(".zarr") for part in Path(filename).parts
    )
```

Reply:
@manzt the reason for that is that there's various parts of the ecosystem that historically haven't played nice with pathlib Paths, see e.g. https://github.com/scikit-image/scikit-image/issues/4138 or https://github.com/scikit-image/scikit-image/issues/3965. I haven't been keeping track of progress community-wide, but basically it felt safer to "contain" the pathlib elements and use strings preferentially except where pathlib is handy.

Perhaps the situation has changed now and we should take the leap and embrace them, but to be honest I've regretted it each time I've taken that leap! 😂

Reply:
Makes sense, thanks for the clarification! To be clear, I wasn't suggesting trying to interact with other parts of ecosystem with pathlib objects, but instead, rather than casting to strings immediately and then trying to glob files or build paths (with `os.path` + strings), just work with pathlib _until_ all desired paths are created.

Essentially wrangle all files that need to be opened with pathlib, and _then_ cast those paths to strings when calling `skimage.imread` etc.